### ISSUE TITLE
dns module issues with powerdns

### ISSUE THREAD

**Issue by defenestration** [prometheus/blackbox_exporter#470]
### Host operating system: output of `uname -a`
Linux c01.b3.blackbox.pro.mon.dev.domain.com 3.10.0-327.36.3.el7.x86_64 #1 SMP Mon Oct 24 16:09:20 UTC 2016 x86_64 x86_64 x86_64 GNU/Linux

### blackbox_exporter version: output of `blackbox_exporter -version`

blackbox_exporter, version 0.14.0 (branch: HEAD, revision: bba7ef76193948a333a5868a1ab38b864f7d968a)
  build user:       root@63d11aa5b6c6
  build date:       20190315-13:32:31
  go version:       go1.11.5

(issue also occurs in v0.13.0)

### What is the blackbox.yml module config.
```---
modules:
  http_2xx:
    prober: http
    http:
  http:
    prober: http
    http:
      valid_status_codes:
      - 100
      - 200
      - 201
      - 202
      - 203
      - 204
      - 205
      - 206
      - 207
      - 208
      - 226
      - 300
      - 301
      - 302
      - 303
      - 304
      - 305
      - 306
      - 307
      - 308
      - 400
      - 401
      - 403
      - 404
      - 500
      - 503
      no_follow_redirects: true
  plesk:
    prober: http
    http:
      valid_status_codes:
      - 100
      - 200
      - 201
      - 202
      - 203
      - 204
      - 205
      - 206
      - 207
      - 208
      - 226
      - 300
      - 301
      - 302
      - 303
      - 304
      - 305
      - 306
      - 307
      - 308
      - 400
      - 401
      - 403
      - 404
      - 500
      no_follow_redirects: true
      tls_config:
        insecure_skip_verify: true
  tcp_4s:
    prober: tcp
    timeout: 4s
  tcp_8s:
    prober: tcp
    timeout: 8s
  tcp_http:
    prober: tcp
    timeout: 8s
    tcp:
      query_response:
      - expect: "^HTTP"
  pop3_banner:
    prober: tcp
    tcp:
      query_response:
      - expect: "^.OK.*"
  pop3s_banner:
    prober: tcp
    tcp:
      query_response:
      - expect: "^.OK.*"
      tls: true
      tls_config:
        insecure_skip_verify: false
  ssh_banner:
    prober: tcp
    tcp:
      query_response:
      - expect: "^SSH"
  dns:
    prober: dns
    dns:
      query_name: www.google.com
      valid_rcodes:
      - NOERROR
      - REFUSED
      - SERVFAIL
    timeout: 6s
  icmp:
    prober: icmp
    timeout: 8s
    icmp:
      preferred_ip_protocol: ip4
  smtp:
    prober: tcp
    timeout: 5s
    tcp:
      query_response:
      - expect: "^220"
```
### What is the prometheus.yml scrape config.
```
- job_name: machine_blackbox_external_discovery
  honor_labels: false
  scrape_interval: 30s
  metrics_path: "/probe"
  file_sd_configs:
  - files:
    - scrape_configs/discovery/*.blackbox_external.json
    refresh_interval: 10m
  relabel_configs:
  - source_labels:
    - __address__
    target_label: __param_target
  - source_labels:
    - __param_target
    target_label: instance
  - source_labels:
    - job
    - __address__
    regex: blackbox_.*_plesk;(.*)
    target_label: __param_target
    replacement: https://$1
    action: replace
  - target_label: __address__
    replacement: c01.b3.blackbox.pro.mon.dev.domain.com:9115
  - target_label: bb_exporter
    replacement: c01.b3.blackbox.pro.mon.dev.domain.com
```
### What logging output did you get from adding `&debug=true` to the probe URL?
```
Logs for the probe:
ts=2019-05-01T21:25:38.124641221Z caller=main.go:118 module=dns target=67.43.6.98:53 level=info msg="Beginning probe" probe=dns timeout_seconds=5.5
ts=2019-05-01T21:25:38.124716122Z caller=utils.go:42 module=dns target=67.43.6.98:53 level=info msg="Resolving target address" preferred_ip_protocol=ip6
ts=2019-05-01T21:25:38.12473614Z caller=utils.go:52 module=dns target=67.43.6.98:53 level=warn msg="Resolution with preferred IP protocol failed, attempting fallback protocol" fallback_protocol=ip4 err="address 67.43.6.98: no suitable address found"
ts=2019-05-01T21:25:38.124752024Z caller=utils.go:65 module=dns target=67.43.6.98:53 level=info msg="Resolved target address" ip=67.43.6.98
ts=2019-05-01T21:25:38.124766356Z caller=dns.go:171 module=dns target=67.43.6.98:53 level=info msg="Making DNS query" target=67.43.6.98:53 dial_protocol=udp4 query=www.google.com type=255
ts=2019-05-01T21:25:38.125304646Z caller=dns.go:176 module=dns target=67.43.6.98:53 level=error msg="Error while sending a DNS query" err="dns: failed to unpack truncated message"
ts=2019-05-01T21:25:38.125330078Z caller=main.go:131 module=dns target=67.43.6.98:53 level=error msg="Probe failed" duration_seconds=0.000644881



Metrics that would have been returned:
# HELP probe_dns_additional_rrs Returns number of entries in the additional resource record list
# TYPE probe_dns_additional_rrs gauge
probe_dns_additional_rrs 0
# HELP probe_dns_answer_rrs Returns number of entries in the answer resource record list
# TYPE probe_dns_answer_rrs gauge
probe_dns_answer_rrs 0
# HELP probe_dns_authority_rrs Returns number of entries in the authority resource record list
# TYPE probe_dns_authority_rrs gauge
probe_dns_authority_rrs 0
# HELP probe_dns_lookup_time_seconds Returns the time taken for probe dns lookup in seconds
# TYPE probe_dns_lookup_time_seconds gauge
probe_dns_lookup_time_seconds 3.049e-05
# HELP probe_duration_seconds Returns how long the probe took to complete in seconds
# TYPE probe_duration_seconds gauge
probe_duration_seconds 0.000644881
# HELP probe_ip_protocol Specifies whether probe ip protocol is IP4 or IP6
# TYPE probe_ip_protocol gauge
probe_ip_protocol 4
# HELP probe_success Displays whether or not the probe was a success
# TYPE probe_success gauge
probe_success 0



Module configuration:
prober: dns
timeout: 6s
dns:
  query_name: www.google.com
  valid_rcodes:
  - NOERROR
  - REFUSED
  - SERVFAIL
```
### What did you do that produced an error?

Blackbox does a DNS check on the server.

Server blackbox is checking is running 'powerdns' on linux

```
[root@test01 ~]# pdnsutil --version
pdnsutil 4.1.5
```

### What did you expect to see?

```probe_success 1```

### What did you see instead?
```probe_success 0```

### Additional 

We have several servers running powerdns (which is installed through CPanel) that this occurs on.  

The server running powerdns returns a 'refused' response when performing a dig or nslookup, which we accept as a valid error code in the blackbox check.  

```
[root@c01.b3.blackbox.pro.mon.dev ~]# nslookup www.google.com. 67.43.6.98
Server:   67.43.6.98
Address:  67.43.6.98#53

** server can't find www.google.com: REFUSED
```

**Comment by brian-brazil** [comment]
This looks like https://github.com/PowerDNS/pdns/issues/6912, the DNS library we use doesn't like the malformed response. I'd suggest upgrading your DNS server.

**Comment by defenestration** [comment]
Thanks, it looks like this should be fixed when they release powerdns 4.2 if i am reading that issue correctly.
